addons/account/wizard/account_advance_payment.py

```diff
@@ -0,0 +1,338 @@
+
+    @api.depends('order_ids')
+    def _compute_display_draft_invoice_warning(self):
+        for wizard in self:
+            wizard.display_draft_invoice_warning = wizard.order_ids.invoice_ids.filtered(lambda invoice: invoice.state == 'draft')
+
+    @api.depends('order_ids')
+    def _compute_invoice_amounts(self):
+        for wizard in self:
+            wizard.amount_invoiced = sum(wizard.order_ids._origin.mapped('amount_invoiced'))
+            wizard.amount_to_invoice = sum(wizard.order_ids._origin.mapped('amount_to_invoice'))
+
+    #=== ONCHANGE METHODS ===#
+
+    @api.onchange('advance_payment_method')
+    def _onchange_advance_payment_method(self):
+        if self.advance_payment_method == 'percentage':
+            amount = self.default_get(['amount']).get('amount')
+            return {'value': {'amount': amount}}
+
+    #=== CONSTRAINT METHODS ===#
+
+    def _check_amount_is_positive(self):
+        for wizard in self:
+            if (wizard.advance_payment_method == 'percentage' and wizard.amount <= 0.00) \
+                    or (wizard.advance_payment_method == 'fixed' and wizard.fixed_amount <= 0.00):
+                raise UserError(_('The value of the down payment amount must be positive.'))
+
+    # === ACTION METHODS ===#
+
+    def create_invoices(self):
+        self._check_amount_is_positive()
+        invoices = self._create_invoices(self.order_ids)
+        return self.order_ids.action_view_invoice(invoices=invoices)
+
+    def view_draft_invoices(self):
+        return {
+            'name': _('Draft Invoices'),
+            'type': 'ir.actions.act_window',
+            'view_mode': 'tree',
+            'views': [(False, 'list'), (False, 'form')],
+            'res_model': 'account.move',
+            'domain': [('state', '=', 'draft')], # To override
+        }
+
+    def _create_invoices(self, orders):
+        if self.advance_payment_method == 'delivered':
+            return orders._create_invoices(final=True, grouped=not self._invoice_grouped())
+
+        return self.create_down_payment_invoice()
+
+    def create_down_payment_invoice(self):
+        self.order_ids.ensure_one()
+        self = self.with_company(self.company_id)
+        order = self.order_ids
+
+        # Create down payment section if necessary
+        OrderLine = self.order_ids._subrecord_type().with_context(no_log_for_new_lines=True)
+        if not any(line.display_type and line.is_downpayment for line in order.order_line):
+            OrderLine.create(
+                self._prepare_down_payment_section_values(order)
+            )
+
+        values, accounts = self._prepare_down_payment_lines_values(order)
+        down_payment_lines = OrderLine.create(values)
+
+        invoice = self.env['account.move'].sudo().create(
+            self._prepare_invoice_values(order, down_payment_lines, accounts)
+        ).with_user(self.env.uid)  # Unsudo the invoice after creation
```

Comment:
Why do we need sudo in the first place ?

Reply:
Added a reference